R/Spectra-functions.R

```diff
@@ -240,50 +248,42 @@ applyProcessing <- function(object, f = dataStorage(object),
 #' fl <- system.file("TripleTOF-SWATH/PestMix1_SWATH.mzML", package = "msdata")
 #' sps <- Spectra(fl, source = MsBackendMzR())
 #'
-#' correlateSpectra <- function(x, y, use = "pairwise.complete.obs", ...) {
-#'     cor(x[, 2], y[, 2], use = use, ...)
-#' }
-#'
 #' sps <- sps[1:10]
-#' res <- .compare_spectra(sps, sps, ppm = 20, FUN = correlateSpectra)
-#' res <- .compare_spectra(sps, sps, FUN = correlateSpectra,
-#'     method = "spearman", ppm = 40)
+#' res <- .compare_spectra_chunk(sps, sps, ppm = 20)
 #'
-#' res <- .compare_spectra(sps[1], sps, FUN = correlateSpectra)
-#' res <- .compare_spectra(sps, sps[1], FUN = correlateSpectra)
+#' res <- .compare_spectra_chunk(sps[1], sps, FUN = correlateSpectra)
+#' res <- .compare_spectra_chunk(sps, sps[1], FUN = correlateSpectra)
 #'
 #' @noRd
-.compare_spectra <- function(x, y = NULL, MAPFUN = joinPeaks, tolerance = 0,
-                             ppm = 20, FUN = ndotproduct, ...) {
-    x_idx <- seq_along(x)
-    y_idx <- seq_along(y)
+.compare_spectra_chunk <- function(x, y = NULL, MAPFUN = joinPeaks,
+                                 tolerance = 0, ppm = 20,
+                                 FUN = ndotproduct, chunkSize = 10000, ...) {
+    nx <- length(x)
+    ny <- length(y)
+    if (nx <= chunkSize && ny <= chunkSize) {
+        mat <- .peaks_compare(.peaksapply(x), .peaksapply(y), MAPFUN = MAPFUN,
+                              tolerance = tolerance, ppm = ppm,
+                              FUN = FUN, ...)
+        dimnames(mat) <- list(spectraNames(x), spectraNames(y))
+        return(mat)
+    }
 
-    nx <- length(x_idx)
-    ny <- length(y_idx)
+    x_idx <- seq_len(nx)
+    x_chunks <- split(x_idx, ceiling(x_idx / chunkSize))
+
+    y_idx <- seq_len(ny)
+    y_chunks <- split(y_idx, ceiling(y_idx / chunkSize))
+    rm(x_idx)
```

Comment:
Just minor coding style point: I would put the `rm(x_idx)` direct below `x_chunks <- split(x_idx, ceiling(x_idx / chunkSize))`. BTW do we need the chunk calculation in other functions as well? If that is the case I would put it in a separate function. Maybe in `MsCoreUtils` if you need this in `Chromatograms` etc.

Reply:
OK, I will move the `rm(x_idx)`.

Since it is a one-liner I would keep the chunk calculation for now within this function. Maybe later - if we indeed will need it also for `Chromatograms` we could define a function for it.